### PR TITLE
Initial Commit

### PR DESCRIPTION
**Pull request by benashwell**

### CODE REVIEW

File: lib/barista.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  attr_accessor :currently_processing_order
+
+  def initialize(orders)
+    @orders = orders
+  end
+
+  def begin_processing_next_order
+    @currently_processing_order = @orders.get_next_pending_order
+    unless currently_processing_order.nil?
+      @orders.update_order_status(@currently_processing_order, "PROCESSING")
```
Comment (agnieszkazbikowska-madetech) [inline lib/barista.rb]: `"PROCESSING"` could be a constant or an enum.

File: lib/barista.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    end
+  end
+
+  def complete_current_order
+    unless @currently_processing_order.nil?
+      puts("Order for " + @currently_processing_order.customers_name + " ready!")
+      @orders.update_order_status(@currently_processing_order, "PROCESSED")
```
Comment (agnieszkazbikowska-madetech) [inline lib/barista.rb]: `"PROCESSED"` could be a constant or an enum.

File: lib/cashier.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  def initialize(orders)
+    @orders = orders
+  end
+
+  def take_customers_order(customers_name)
+    if customers_name.nil? || customers_name.empty?
+      raise OrderError.new
+    end
+
+    @orders.add_order(Order.new(customers_name))
```
Comment (agnieszkazbikowska-madetech) [inline lib/cashier.rb:16]: Is it possible for 1 customer to have more than order? If this is the case we might need something more unique than `customers_name` for the order name. If not, we can add a check if a given customer has already placed an order.

File: lib/order.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+class Order
```
Comment (agnieszkazbikowska-madetech) [inline lib/order.rb:1]: I don't think that there are tests for this file.

File: lib/order.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+  attr_accessor :customers_name, :status
+
+  def initialize(name, status = "PENDING")
```
Comment (agnieszkazbikowska-madetech) [inline lib/order.rb]: `"PENDING"` could be a constant or an enum.

File: lib/orders.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
+require "order"
```
Comment (agnieszkazbikowska-madetech) [inline lib/orders.rb:1]: I don't think that there are tests for this file.

File: lib/orders.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+class Orders
+  attr_reader :orders
+
+  def initialize
+    @orders = []
+  end
+
+  def add_order(order)
+    @orders << order
```
Comment (agnieszkazbikowska-madetech) [inline lib/orders.rb:17]: Is it possible for the `order` to be empty?

File: lib/orders.rb [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  end
+
+  def get_processing_orders
+    return get_orders_by_status("PROCESSING")
```
Comment (agnieszkazbikowska-madetech) [inline lib/orders.rb]: `"PROCESSING"` could be a constant or an enum.

File: lib/orders.rb [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  end
+
+  def get_processed_orders
+    return get_orders_by_status("PROCESSED")
```
Comment (agnieszkazbikowska-madetech) [inline lib/orders.rb]: `"PROCESSED"` could be a constant or an enum.

File: lib/orders.rb [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+  end
+
+  def get_orders_by_status(status)
```
Comment (agnieszkazbikowska-madetech) [inline lib/orders.rb]: If there is function `get_orders_by_status` do we need separate `get_processing_orders` and `get_processed_orders`? Whenever this is called, we could just call `get_orders_by_status` directly.

File: lib/orders.rb [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    @orders.each do |order|
+      if order.status == status
+        processing_orders << order
+      end
+    end
+    return processing_orders
+  end
+
+  def get_next_pending_order
+    return @orders.detect { |order| order.status == "PENDING" }
```
Comment (agnieszkazbikowska-madetech) [inline lib/orders.rb]: `"PROCESSING"` could be a constant or an enum.

File: spec/Barista_spec.rb [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+
+  context "only processed orders in queue" do
+    it "processes no orders" do
+      @orders.add_order(Order.new("Ben", "PROCESSED"))
+      @barista.begin_processing_next_order
+      expect(@barista.currently_processing_order).to be_nil
+    end
+  end
+
+  context "pending order are moved to processing" do
```
Comment (agnieszkazbikowska-madetech) [inline spec/Barista_spec.rb]: `pending orders are moved to processing`?

File: spec/Barista_spec.rb [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+  end
+
+  context "No orders have come in" do
+    it "processes no orders" do
+      @barista.begin_processing_next_order
+      expect(@barista.currently_processing_order).to be_nil
+    end
+  end
+
+  context "only processed orders in queue" do
```
Comment (agnieszkazbikowska-madetech) [inline spec/Barista_spec.rb]: `only processed orders are in queue`?

File: spec/Barista_spec.rb [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+    it "processes no orders" do
+      @orders.add_order(Order.new("Ben", "PROCESSED"))
+      @barista.begin_processing_next_order
+      expect(@barista.currently_processing_order).to be_nil
+    end
+  end
+
+  context "pending order are moved to processing" do
+    it "orders is being processed" do
```
Comment (agnieszkazbikowska-madetech) [inline spec/Barista_spec.rb]: `orders are being processed`?

File: spec/Barista_spec.rb [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+  context "No current order processing" do
+    it "processes no orders" do
+      @barista.complete_current_order
+      expect(@barista.currently_processing_order).to be_nil
+      expect(@orders.get_processed_orders.length).to eq(0)
+    end
+  end
+
+  context "current order processed" do
+    it "calls customer by name" do
```
Comment (agnieszkazbikowska-madetech) [inline spec/Barista_spec.rb]: `calls the customer by name`?

**Review by agnieszkazbikowska-madetech — APPROVED** [review]
This looks really good! I really like the clear naming convention, good test coverage, clear project structure and clear separation of responsibilities (one function does one thing). I've added a few small comments here and there but they are really small. What I could advise is to add a readme, it would help running & testing the project.

File: lib/cashier.rb (hunk 3 above)
Comment (benashwell) [reply lib/cashier.rb:16]: Agreed, currently it is designed for a single order by name but I have further ideas on having multiple drinks per customer and then having an order ID which would be returned to the customer.

File: lib/order.rb (hunk 4 above)
Comment (benashwell) [reply lib/order.rb:1]: There is not, this was created as part of refactoring from the Cashier class so it should be covered by tests in the Cashier spec

File: lib/orders.rb (hunk 6 above)
Comment (benashwell) [reply lib/orders.rb:1]: There is not, this was created as part of refactoring from the Cashier class so it should be covered by tests in the Cashier spec

File: lib/orders.rb (hunk 7 above)
Comment (benashwell) [reply lib/orders.rb:17]: not currently, but you are right nil protection here is probably a good idea

File: lib/orders.rb (hunk 10 above)
Comment (benashwell) [reply lib/orders.rb]: We could, however I found that it reads poorly in the other methods. My intent is get_orders_by_status to be a private method so no one outside of the orders class can access it.

**Pull request merged**